### PR TITLE
Manages Database Connection on Startup

### PR DESCRIPTION
**Pull request by JulesRoadknight**
- Uses regular naming conventions
- Removes unnecessary casting
- Database connection automated
- Improves Spy security

### CODE REVIEW

File: src/main/java/Constants.java [hunk 1]
```diff
@@ -1,26 +1,41 @@
 public final class Constants {
 
-    private Constants() {}
-
-    public static final String menuOptions = "Please select an option:\n" +
+    public static final String MENUOPTIONS = "Please select an option:\n" +
```
Comment (sohbaker) [inline src/main/java/Constants.java:3]: Like the use of capitals for Constants here. 

Naming conventions use capitals and an underscore between words: https://www.oracle.com/technetwork/articles/javaee/codeconventions-135099.html

File: src/main/java/Constants.java [hunk 2]
```diff
@@ -1,26 +1,41 @@
...
             "6. Email";
 
-    public static final String prodContactManagerDB = "jdbc:postgresql://localhost:5432/contactmanagerdb";
+    public static final String LOCALCONNECTION = "jdbc:postgresql://localhost:5432/";
+
+    public static final String CREATEDB = "CREATE DATABASE prod_contact_manager";
+
+    public static final String CREATETESTDB = "CREATE DATABASE test_contact_manager";
+
+    public static final String PRODDATABASE = "prod_contact_manager";
```
Comment (sohbaker) [inline src/main/java/Constants.java:25]: Do we use these strings separately from the `CREATEDB` and `CREATETESTDB` strings? Wondering whether you have some duplication. 

If these strings are needed seperately, is there a way to interpolate the names into the `CREATE DATABASE` sql statement?

File: src/main/java/Database.java [hunk 3]
```diff
@@ -1,24 +1,19 @@
...
+    public Database(ConsoleIO consoleIO, Connection connection) {
         this.consoleIO = consoleIO;
-        this.dbName = dbName;
-        try {
-            Class.forName("org.postgresql.Driver");
-            connection = DriverManager.getConnection(databaseConnection, "postgres", "contactManager1");
-        } catch (Exception e) {
-            e.printStackTrace();
-        }
+        this.tableName = "contactmanagerdb";
```
Comment (sohbaker) [inline src/main/java/Database.java:15]: Hardcoded string here. Could you do this in a different way?

File: src/main/java/Database.java [hunk 4]
```diff
@@ -34,34 +29,48 @@ public static String getDBColumnName(int field) {
...
-            statement = connection.createStatement();
-            String addContact = "INSERT INTO " + dbName + " VALUES(DEFAULT, '" +
-                    contact.getFirstName() + "', '" + contact.getLastName() + "', '" + contact.getAddress() + "', '" + contact.getPhoneNumber() + "', '" + contact.getDOB() + "', '" + contact.getEmail() +
-                    "');";
-            statement.execute(addContact);
-            statement.close();
+            parsed = format.parse(contact.getdOB());
+        } catch (ParseException e) {
+            throw new RuntimeException(e);
+            // this is giant faff that needs fixing somewhere else...
```
Comment (sohbaker) [inline src/main/java/Database.java:39]: Uncommented code, but like the inclination expressed here.

File: src/main/java/Database.java [hunk 5]
```diff
@@ -84,28 +94,32 @@ public Contact getContact(int index) throws SQLException {
                     setContact.getString("email"),
                     consoleIO
             );
-        statement.close();
-        return contact;
+            statement.close();
+            return contact;
+        } else {
+            return null; // <-- this is going to catch you out sooooooooooooo bad. There is a type! Option<Contact>
```
Comment (sohbaker) [inline src/main/java/Database.java:100]: Like the inclination here too. How might you control this?

File: src/test/java/ContactListSpy.java [hunk 6]
```diff
@@ -47,4 +48,29 @@ public Contact getContact(int index) {
         this.getContactHasBeenCalled = true;
         return null;
     }
+
+    @Override
+    public List<Contact> getContacts() {
+        return contactArray;
+    }
+
+    public boolean returnCreateContactHasBeenCalled() {
```
Comment (sohbaker) [inline src/test/java/ContactListSpy.java:57]: Nice!

File: src/test/java/DatabaseTest.java [hunk 7]
```diff
@@ -7,17 +7,18 @@
...
-    Statement statement;
-    ArrayList<Contact> contactArray;
+    private Database database;
+    private Contact contact;
+    private Contact secondContact;
+    private Connection connection;
+    private Statement statement;
+    public ArrayList<Contact> contactArray;
 
     public static String defaultFirstName = "Jamey";
```
Comment (sohbaker) [inline src/test/java/DatabaseTest.java:23]: I think these potentially could be made private as well?

File: src/main/java/Contact.java [hunk 8]
```diff
@@ -2,98 +2,87 @@
...
-    public static final String DOBField = "Date of Birth (dd/MM/yyyy)";
-    public static final String EmailField = "Unique Email Address";
+    private String firstName;
+    private String lastName;
+    private String address;
+    private String phoneNumber;
+    private String dOB;
+    private String email;
+
+    public static final String firstNameField = "First Name";
```
Comment (nickbdyer) [inline src/main/java/Contact.java:12]: Not sure these need to be `static`?

File: src/main/java/Contact.java [hunk 9]
```diff
@@ -2,98 +2,87 @@
...
-            case 3: return AddressField;
-            case 4: return PhoneNumberField;
-            case 5: return DOBField;
-            default: return EmailField;
+            case 1: return firstNameField;
+            case 2: return lastNameField;
+            case 3: return addressField;
+            case 4: return phoneNumberField;
+            case 5: return dOBField;
+            default: return emailField;
```
Comment (nickbdyer) [inline src/main/java/Contact.java:60]: This is a surprising `default` setting. 

Maybe default should return `null` or something less specific than the email field.

File: src/main/java/Database.java [hunk 10]
```diff
@@ -1,24 +1,19 @@
 import java.sql.*;
-import java.util.ArrayList;
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.LinkedList;
```
Comment (nickbdyer) [inline src/main/java/Database.java:4]: Exciting

File: src/main/java/Database.java [hunk 11]
```diff
@@ -34,34 +29,48 @@ public static String getDBColumnName(int field) {
...
+            create.setString(1, contact.getFirstName());
+            create.setString(2, contact.getLastName());
+            create.setString(3, contact.getAddress());
+            create.setString(4, contact.getPhoneNumber());
+            create.setDate(5, new Date(parsed.getTime()));
+            create.setString(6, contact.getEmail());
+            create.execute();
+            create.close();
         } catch (SQLException e) {
+            System.out.println(e);
```
Comment (nickbdyer) [inline src/main/java/Database.java:53]: Hmmmmm.

File: src/main/java/Run.java [hunk 12]
```diff
@@ -1,14 +1,107 @@
...
 
+import static java.lang.System.exit;
+
 public class Run {
+
     public static void main(String[] args) {
         ConsoleIO consoleIO = new ConsoleIO(System.in, System.out);
+
+        startPostgres();
+        startDatabaseCluster();
```
Comment (nickbdyer) [inline src/main/java/Run.java:16]: Somewhat misleading if you there's only one database.

File: src/main/java/Run.java [hunk 13]
```diff
@@ -1,14 +1,107 @@
...
+        Connection connection = connectLocalhost();
+
+        createDatabase(connection, createDB);
+        connection = connectDatabase(connection, Constants.LOCALCONNECTION, databaseName);
+        createTable(connection);
+
+        return connection;
+    }
+
+    public static Connection connectLocalhost() {
```
Comment (nickbdyer) [inline src/main/java/Run.java:60]: This is where abstracting out environment variables can avoid methods like this that only apply in certain circumstances. Your database will rarely be available locally.

**Review by nickbdyer — COMMENTED** [review]
Feels like there is room for some integration/acceptance tests here, that validate that things actually talk to each other without mocks.

**Pull request merged**